ballerina-tests/tests/service_dispatching_header_param_binding_ideal_mapping_test.bal

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/test;
+
+listener http:Listener HeaderBindingIdealEP = new(headerParamBindingIdealTestPort);
+http:Client headerBindingIdealClient = check new("http://localhost:" + headerParamBindingIdealTestPort.toString());
+
+@http:ServiceConfig {
+    treatNilableAsOptional : false
+}
+service /headerparamservice on HeaderBindingIdealEP {
+
+    resource function get test1(@http:Header string foo, int bar) returns json {
+        json responseJson = { value1: foo, value2: bar};
+        return responseJson;
+    }
+
+    resource function get test2(@http:Header string? foo, int bar) returns json {
+        json responseJson = { value1: foo ?: "empty", value2: bar};
+        return responseJson;
+    }
+}
+
+@test:Config {}
+function testIdealHeaderParamBindingWithHeaderValue() {
+    http:Response|error response = headerBindingIdealClient->get("/headerparamservice/test1?foo=WSO2&bar=56", {"foo":"Ballerina"});
```

Comment:
Can't we use client data-binding here ? Check for other places as well.

Reply:
Resolved in [`dfed505`](https://github.com/ballerina-platform/module-ballerina-http/pull/653/commits/dfed50595eb141f7941d3405f50e6e299374c288)